### PR TITLE
docs(codepipeline): Added documentation for creating custom actions

### LINKED ISSUE #11029: [@aws-cdk/aws-codepipeline-actions] use a custom build action in CDK project
https://docs.aws.amazon.com/cdk/api/latest/docs/aws-codepipeline-actions-readme.html

I am trying to figure out how to reference a custom action for a build stage within my CDK template. I have spent a few hours reading the documentation and source code and I can't see a way to use a custom build action in my pipeline. I have created the pipeline via the console and I am able to use the custom action without any issues.

The custom action I am trying to utilize was actually created from the cloudformation template in this project https://github.com/aws-samples/aws-codepipeline-custom-action

Hope I am just missing something and there is a way with the latest version of the CDK to utilize the action.

---

This is a 📕 documentation issue

### PR DESCRIPTION
**Pull request by phillipCouto**
Add a reference to the codepipeline README to implement the IAction interface for custom actions.

Added a code example on IAction to serve as a starting point for implementing the type.

fixes #11029

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-codepipeline/lib/action.ts [hunk 1]
```diff
@@ -119,7 +119,79 @@ export interface ActionConfig {
 }
 
 /**
- * A Pipeline Action
+ * A Pipeline Action.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/lib/action.ts:122]: Can you move it to the ReadMe file? I think this is too complicated for the JSDocs of an interface...

**Review by skinny85 — CHANGES_REQUESTED** [review]
Thanks for the contribution @phillipCouto !

Before I dive deeply into the content, I have one starting comment 🙂.

File: packages/@aws-cdk/aws-codepipeline/lib/action.ts (hunk 1 above)
Comment (phillipCouto) [reply packages/@aws-cdk/aws-codepipeline/lib/action.ts:122]: Sure thing :)

Comment (phillipCouto) [reply packages/@aws-cdk/aws-codepipeline/lib/action.ts]: @skinny85 check now.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Hey @phillipCouto , quick question before I dive deeper here.

When you say "custom actions", do you mean an Action for something that's supported in CodePipeline, but not yet in the CDK? Or is it "custom" in the sense of [CodePipeline custom actions](https://docs.aws.amazon.com/codepipeline/latest/userguide/actions-create-custom-action.html)?

If it's the former, we need to change the wording here a little bit (maybe something like "adding support for new actions"?). if it's the latter, then this is much more complicated - see the [`JenkinsProvider` of the `JenkinsAction`](https://github.com/aws/aws-cdk/blob/e21a1d9c369748ebf7081128b0dbdb76677cb6c4/packages/%40aws-cdk/aws-codepipeline-actions/lib/jenkins/jenkins-provider.ts#L137-L137).

**Comment by phillipCouto** [comment]
> Hey @phillipCouto , quick question before I dive deeper here.
> 
> When you say "custom actions", do you mean an Action for something that's supported in CodePipeline, but not yet in the CDK? Or is it "custom" in the sense of [CodePipeline custom actions](https://docs.aws.amazon.com/codepipeline/latest/userguide/actions-create-custom-action.html)?
> 
> If it's the former, we need to change the wording here a little bit (maybe something like "adding support for new actions"?). if it's the latter, then this is much more complicated - see the [`JenkinsProvider` of the `JenkinsAction`](https://github.com/aws/aws-cdk/blob/e21a1d9c369748ebf7081128b0dbdb76677cb6c4/packages/%40aws-cdk/aws-codepipeline-actions/lib/jenkins/jenkins-provider.ts#L137-L137).

So reviewing the Jenkins code you referenced and comparing to what I was documenting I guess what I am trying to document is using a custom action that has been registered in the AWS Account outside of the CDK stack. The Jenkins code registers the custom action and then allows you to use it in your pipeline.

If my explanation makes sense and you are happy with me just documenting how to use a custom action that is already registered then I will rework my changes with that scope.

File: packages/@aws-cdk/aws-codepipeline/README.md [hunk 2]
```diff
@@ -98,6 +98,87 @@ or you can use the `IStage.addAction()` method to mutate an existing Stage:
...
+        return {
+            // The configuration expected by the custom action.
+            configuration: {
+                ...
+            }
+        };
+    }
+
+    onStateChange(name: string, target?: events.IRuleTarget, options?: events.RuleProps): events.Rule {
+        const rule = new events.Rule(this.scope, name, options);
```
Comment (umang94) [inline packages/@aws-cdk/aws-codepipeline/README.md:147]: Shouldn't this be 
`` const rule = new events.Rule(this._scope, name, options);`` ?

Comment (phillipCouto) [reply packages/@aws-cdk/aws-codepipeline/README.md:147]: We are using `this.scope` as there is a getter defined further down that will throw a descriptive error if the `onStateChange` is called before `bind`. This is the same for `this.stage`.

**Review by umang94 — COMMENTED** [review]
Thanks a lot for adding this documentation. I am currently working on a project requiring custom actions and your commit is the only documentation I am able to refer to. Hence the feedback as well.

File: packages/@aws-cdk/aws-codepipeline/README.md (hunk 2 above)
Comment (umang94) [reply packages/@aws-cdk/aws-codepipeline/README.md:147]: I see. Thank you for the clarification. I am not adequately verse with the Typescript magic. I have the deleted the other similar comments I had.

File: packages/@aws-cdk/aws-codepipeline/README.md [hunk 3]
```diff
@@ -98,6 +98,87 @@ or you can use the `IStage.addAction()` method to mutate an existing Stage:
...
+        this._scope = scope;
+        return {
+            // The configuration expected by the custom action.
+            configuration: {
+                ...
+            }
+        };
+    }
+
+    onStateChange(name: string, target?: events.IRuleTarget, options?: events.RuleProps): events.Rule {
```
[the remark this answers is not in the document]
Comment (phillipCouto) [reply packages/@aws-cdk/aws-codepipeline/README.md:146]: Unfortunately it is required by the [IAction](https://docs.aws.amazon.com/cdk/api/latest/typescript/api/aws-codepipeline/iaction.html#aws_codepipeline_IAction_onStateChange) interface. So for now you can just use the code I have included and then maybe a change to `IAction` in the future could make this optional so only custom actions that require an event rule for state change would implement it.

I will add a comment stating that in the example.

File: packages/@aws-cdk/aws-codepipeline/README.md (hunk 2 above)
Comment (phillipCouto) [reply packages/@aws-cdk/aws-codepipeline/README.md:147]: Not a problem, I actually reviewed the code to make sure I still needed it 😆.

**Comment by skinny85** [comment]
@phillipCouto I recently made a change to introduce an `Action` class to the `codepipeline` module that contains some common logic. Any chance you could update your PR to extend this new class? I think it will simplify the logic of implementing a custom Action considerably.

Let me know!

**Comment by skinny85** [comment]
Class is here: https://github.com/aws/aws-cdk/blob/61ad0aa84a98aa4db1b94adcd6fdc2cd10e908d5/packages/%40aws-cdk/aws-codepipeline/lib/action.ts#L267-L267 (will be released as part of CDK version `1.99.0`).

**Review by skinny85 — DISMISSED** [review]
Putting this in 'Request changes' to clear it from my ToDo list, @phillipCouto let me know if you're still working on this!

**Comment by phillipCouto** [comment]
> Putting this in 'Request changes' to clear it from my ToDo list, @phillipCouto let me know if you're still working on this!

My apologizes I kept forgetting to come back to it. I am going to work on it tonight.

**Comment by phillipCouto** [comment]
> @phillipCouto I recently made a change to introduce an `Action` class to the `codepipeline` module that contains some common logic. Any chance you could update your PR to extend this new class? I think it will simplify the logic of implementing a custom Action considerably.
> 
> Let me know!

Much better!!! Thank you for that!

Now I need to go and upgrade all of our stacks that use the old IAction code 😆.

**Pull request closed without merging**